tensorflow/core/kernels/non_max_suppression_op.cu.cc

```diff
@@ -227,91 +208,111 @@ __global__ void Iota(const int num_elements, const T offset, T* to_fill) {
   }
 }
 
-tensorflow::Status NmsGpu(const float* d_desc_sorted_boxes_float_ptr,
+tensorflow::Status NmsGpu(const float* d_sorted_boxes_float_ptr,
                           const int num_boxes, const float thresh,
-                          int* d_keep_sorted_list, int* h_nkeep,
-                          int* dev_delete_mask, int* host_delete_mask,
+                          int* d_selected_indices, int* h_nkeep,
                           OpKernelContext* context, bool flip_boxes) {
-  // d_desc_sorted_boxes_float_ptr is a pointer
-  //    to device memory float array containing the box corners for N boxes.
-  // N is number of boxes.
-  // threshold is the intersection-over-union threshold for elimination
-  // d_keep_sorted_list is a device pointer to int array containing sorted
-  //    indices of the boxes to keep h_nkeep is the pointer to number of
-  //    elements to keep for the host.
-  // h_nkeep is a host pointer for returning number of items to keep.
-  // flip_boxes flag reorders the boxes use lower left and upper right corners
-  // if they are given in mixed format. Making sure we respect the __align(16)__
+  // Making sure we respect the __align(16)__
   // we promised to the compiler.
-  auto iptr = reinterpret_cast<std::uintptr_t>(d_desc_sorted_boxes_float_ptr);
-  CHECK_EQ((iptr & 15), 0);
+  auto iptr = reinterpret_cast<std::uintptr_t>(d_sorted_boxes_float_ptr);
+  if ((iptr & 15) != 0) {
+    return errors::InvalidArgument("Boxes should be aligned to 16 Bytes.");
+  }
+  // allocate bitmask arrays on host and on device
+  Tensor h_nms_mask, d_nms_mask;
+  const int bit_mask_len =
+      (num_boxes + kNmsBoxesPerThread - 1) / kNmsBoxesPerThread;
+
+  int64 max_nms_mask_size = num_boxes * bit_mask_len;
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_INT32, TensorShape({max_nms_mask_size}), &d_nms_mask));
+  // reset data sensitive tensors
+  auto device = context->eigen_gpu_device();
+  auto config = GetCudaLaunchConfig(d_nms_mask.NumElements(), device);
+  TF_CHECK_OK(GpuLaunchKernel(SetZero<int>, config.block_count,
+                              config.thread_per_block, 0, device.stream(),
+                              config.virtual_thread_count,
+                              d_nms_mask.flat<int32>().data()));
 
-  // The next kernel expects squares
+  AllocatorAttributes alloc_attr;
+  alloc_attr.set_on_host(true);
+  alloc_attr.set_gpu_compatible(true);
+  TF_RETURN_IF_ERROR(context->allocate_temp(DataType::DT_INT32,
+                                            TensorShape({max_nms_mask_size}),
```

Comment:
It look like we only need `num_to_copy*2` elements in this buffer, i.e. a circular array. Please either fix it or add comment for the future reference/optimization.

Reply:
I will do that. But very skeptical about any benefits such an optimization would bring. Typical ranges of num_boxes in the wild is around 2k to 6k, that would make that host buffer in the order of 6k x 6k/8=4.5MB. for a chunk size of 2000, circular buffer you suggested is 4kx6k/8=3MB . I believe we can save much more memory through other means, not necessarily in this kernel.